### ISSUE TITLE
PCP shutdown stuck with certain BPF/BCC modules

### ISSUE THREAD

**Issue by myllynen** [iovisor/bcc#1852]
With PCP BCC PMDA plugin it seems that certain combinations of BPF/BCC modules cause the related PCP processes not to exit cleanly but get stuck.

Since I can't see anything that could explain this on the PCP side (in BCC PMDA modules nor BPF programs) I hope this could be investigated on the BCC (or even BPF/kernel) side. There was earlier an issue with similar symptoms (see https://github.com/iovisor/bcc/pull/1549) but since here the modules most easily triggering this are not the same the root cause here might be entirely different.

Steps to produce:

1. Install Fedora 28 + updates + kernel 4.17 from updates-testing, reboot
2. `dnf --enablerepo=updates-testing install pcp-pmda-bcc` (will give PCP 4.1 + BCC 0.6.0)
3. `setenforce 0` (just in case)
4. Set `modules = biolatency,biotop,runqlat` in _/var/lib/pcp/pmdas/bcc/bcc.conf_
5. `systemctl stop pmcd pmlogger pmie pmproxy`
6. `systemctl start pmcd`
7. `cd /var/lib/pcp/pmdas/bcc`
8. `./Install`
9. Generate some system activity (like `find /` or `pminfo -f`) so that before the next step `pminfo -f bcc` shows values for all modules
10. `systemctl stop pmcd`

The last step will take only a second or so if using PCP BCC PMDA module like sysfork but at least with these three modules it's most often the case that the command takes ~half a minute (at which point the shell script wrapper will kill the stuck processes forcefully) and the Python process implementing the PCP BCC PMDA plugin being in defunct state.

In case you'd need some more details how to reproduce this, just let me know. Thanks.

**Comment by palmtenor** [comment]
I can indeed repro this, and seems it only happens on the exact combination of `biolatency,biotop,runqlat`, not any one or two of them... I tried to `perf record` the `systemctl` call and most of the stuck time are spent on `nanosleep`, and if you `strace` the `systemctl` call it seems one of the `recvmsg` call returned `-EAGAIN` and it got stuck for a while before doing another `ppoll` and exit.

Maybe it's waiting on something that is not ready yet, and there is a artificial delay before retrying? I don't have too much knowledge on how `systemd` does the shutdown:( Also, how does PCP run the BCC tools? I don't see it runs either as a separated Process or a Thread within `pmcd`?

**Comment by myllynen** [comment]
Thanks for looking into this! Even when using systemd, `pmcd.service` just uses `/usr/share/pcp/lib/pmcd` to start/stop the `pmcd` daemon and it seems that when updating `pmcd.service` to start the daemon directly this is not happening anymore:

```
...
[Service]
Type=forking
EnvironmentFile=/etc/pcp.conf
ExecStart=/usr/libexec/pcp/bin/pmcd
...
```

However, with this in place the PCP BCC PMDA logs are missing (nothing in `/var/log/pcp/pmcd/bcc.log`). PCP currently bundles its own (sometimes slightly modified) copies of BCC tools that are started by the Python BCC PMDA which in turn is started by `pmcd` so there are indeed quite a few moving pieces in the puzzle.

@natoscott Would you have any insight on this? Why do BCC PMDA logs go missing if not using the wrapper script to start/stop `pmcd` and could the use of the wrapper have something to do with these hangs? Thanks.

**Comment by myllynen** [comment]
I investigated this a bit more and found something, perhaps it's even helpful. The above `pmcd.service` change was incomplete as logs were missing (they were found under `/`), this would be complete and correct change for `pmcd.service` that also allows the PMDA plugins to log and close down properly:

```
...
[Service]
Type=forking
Restart=always
EnvironmentFile=/etc/pcp.conf
EnvironmentFile=-/etc/sysconfig/pmcd
WorkingDirectory=/var/log/pcp/pmcd
ExecStart=/usr/libexec/pcp/bin/pmcd
KillMode=mixed
TimeoutSec=15
...
```

But when using this unit file, we see the BCC PMDA Python process being defunct again, just like with the shell script earlier. If we comment out the `KillMode` directive, then the `pmcd` service termination is instant, but PMDA plugins don't have time to clean up properly (nothing is written into their logs during shutdown in case no `KillMode` specified - https://www.mankier.com/5/systemd.kill describes the differences).

**Comment by natoscott** [comment]
@myllynen this turned out to be a bug in the PCP pmcd(1) daemon.  Fixed now (by Mark Goodwin) and will be included in pcp-4.1.1.

**Comment by palmtenor** [comment]
@myllynen Is the problem solved with the `pmcd` fix?

**Comment by natoscott** [comment]
@palmtenor I believe Marko is offline for a few weeks.  I can confirm the pmcd fix will be in pcp-4.1.1 though and I believe this issue could now be closed.
